### DISCUSSION TITLE
A few questions regarding clustering key and compaction

### DISCUSSION THREAD

**Discussion by aping** [milvus-io/milvus#38050] [Q&A and General discussion]
So I'm been trying clustering compaction and want to get a better understanding of it.

1. are segments grouped by partitionID and channel?
I see [here](https://github.com/milvus-io/milvus/blob/v2.4.15/internal/datacoord/meta.go#L352).
is channel equivalent to shard? I mean if I have 100 partitions, 2 shards, does it mean it'll have 200 groups?

2.  and if I understand correctly, within each group (partition+channel), buckets are calculated and rows go into corresponding bucket according to the clustering key?

3. about manual and auto clustering compaction.
if I trigger a manual compaction, or enable auto clustering compaction (and after `dataCoord.compaction.clustering.maxInterval`), it seems that all segments are rewritten and reindexed? even if they don't change at all? I hope I was wrong because it's really heavy for large collections

**Comment by yhmo** [comment]
This discussion is about the shard/channel:
https://github.com/milvus-io/milvus/discussions/37594#discussioncomment-11220675

![channels](https://github.com/user-attachments/assets/b9409e7d-1f9b-4e0b-8866-6bdf845a3fa2)

For channel we have physical channel and virtual channel, the number of physical channels is equal to shard number.
If a collection has 100 partitions, shards=2, there will be at least 400 virtual channels. Each virtual channel consumes data from Pulsar/Kafka and generates segments. Yes, we can say there are 400 "groups" of segments.

Clustering key re-org the segments within a "group" to boost search performance: https://milvus.io/docs/clustering-compaction.md#Clustering-Compaction

Yes, clustering compaction is a heavy operation since it re-org the segments within each "group". Before clustering compaction, all the data in a group is randomly distributed in different segments. After clustering compaction, all the data is sorted by clustering key and sequentially stored into a new batch of segments.

**Reply by aping** [reply]
> Yes, clustering compaction is a heavy operation since it re-org the segments within each "group". Before clustering compaction, all the data in a group is randomly distributed in different segments. After clustering compaction, all the data is sorted by clustering key and sequentially stored into a new batch of segments.

no I mean even when the collection is already clustering compacted and in good shape. for example I did one clustering compaction for a collection, when it finished, I triggered another manual clustering compaction, when there's no inserts or deletes in between, I still see plans generated and it seems that all segments are rewritten and reindexed.

and from [here](https://github.com/milvus-io/milvus/blob/v2.4.15/internal/datacoord/compaction_policy_clustering.go#L151) and [here](https://github.com/milvus-io/milvus/blob/v2.4.15/internal/datacoord/compaction_policy_clustering.go#L240), it seems no difference between a manual compaction and auto compaction when `dataCoord.compaction.clustering.maxInterval` has passed.

so question is, even if the collection is in perfect shape, if I trigger a manual clustering compaction, or enable auto clustering compaction, is all data still going to be rewritten and reindexed?

**Reply by xiaofan-luan** [reply]
Yes, usually you don't need to trigger manual compaction.
But once manual compaction is triggered, it will be executed and data will be reindexed.

I would recommend u to set a periodic clustering compaction or just tune the parameter of when the cluster compaction should be triggered

**Reply by aping** [reply]
in that case I dont think I should enable auto clustering compaction? because otherwise every `dataCoord.compaction.clustering.maxInterval` the collection is rewritten?

so here's my current situation, we have a 5B rows collection, with a scalar column as both partition key and clustering key:
- every day we write ~15M new rows to the collection, with ~500 unique clustering keys
- every once in a while we delete ~15M rows from the collection, with ~500 unique clustering keys
- I'm hoping to delete ~1B rows from it but it seems very expensive that's why I'm trying with clustering key

so question is what happens if I do not enable auto clustering compaction? I mean during inserts/deletes, mixed compactions still run, do they honor the clustering key?

> I would recommend u to set a periodic clustering compaction or just tune the parameter of when the cluster compaction should be triggered

can you elaborate on this? I know I should increase `dataCoord.compaction.clustering.timeout` to 86400 or something based on the size of our collection, plus increasing `dataCoord.compaction.taskQueueCapacity` because the default value 256 is just too small and I see in newer versions (I'm at 2.4.15) it's increased to 100000, anything else I should be aware of?

**Comment by xiaofan-luan** [comment]
clustering compaction won't help on delete.
Delete so much data might badly hurt the cluster.
if you can use partition and drop one partition at a time that might help

**Comment by xiaofan-luan** [comment]
clustering compaction only help on search performance. please contact me on james.luan@zilliz.com and we'd glad to know more about your use case and figure out a way to help
